### PR TITLE
Remove Cast before and after Gelu

### PR DESCRIPTION
**Pull request by iK1D**
Some mix-precision models will cast data to float before Gelu and cast back to half after. This PR changes the Gelu and GeluGrad CUDA kernel to use float for compute internally, and remove these Cast nodes from graph on CUDA EP.
- Change Gelu/GeluGrad CUDA kernel to use float internally will not harm the perf, while we can have better precision.
- This PR doesn't handle ROCm EP as currently we don't have Gelu/GeluGrad ROCm kernel.
- This PR doesn't handle FastGelu, as the kernel implementation for FastGelu is to complicated to changed to use float internally.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/activation/activations_impl.cu [hunk 1]
```diff
@@ -43,6 +43,13 @@ struct OP_Gelu : public CtxGelu {
   }
 };
 
+template <>
+struct OP_Gelu<half> : public CtxGelu {
+  __device__ __inline__ half operator()(const half& a) const {
+    return static_cast<half>(_Gelu(static_cast<float>(a)));
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/activation/activations_impl.cu:49]: Change the Gelu half kernel will impact performance (it adds two data type conversion, even though latency impact might be small). 

If user prefer better accuracy, they will use fp32 version of Gelu anyway. If they need performance, they would prefer the original implementation.

**Comment by tianleiwu** [comment]
Is there a way to generalize the fusion for element-wise operator, like 
-->Element-wise Op1 --> Element-wise Op2--> can be fused as Element-wise Op1(Op2(x))

**Comment by ytaous** [comment]
Wouldn't the Cast Propagation feature (https://github.com/microsoft/onnxruntime/pull/7454) that Satya implemented already covers this case?

File: onnxruntime/contrib_ops/cuda/activation/activations_impl.cu (hunk 1 above)
Comment (iK1D) [reply onnxruntime/contrib_ops/cuda/activation/activations_impl.cu:49]: As mentioned in the PR description, the perf impact is very small from Nsight Compute. The kernel for Gelu and GeluGrad will call _Normcdf and _Exp, and actually the half compute for these 2 functions a also cast data to float for compute and then cast back, so the cast is there anyway. Below are the comparison from Nsight Compute (baseline is the old impl):

Gelu:
image

GeluGrad:
image

**Comment by iK1D** [comment]
> Wouldn't the Cast Propagation feature (#7454) that Satya implemented already covers this case?

Right. Previously I just want to enabled this for CUDA only, but the PropagateCastOps cannot do this. But on second thought, we don't have MLFloat16 kernel for CPU for Gelu/GeluGrad, it's safe to use PropagateCastOps. I've changed the code to use PropagateCastOps by promoting Gelu from Level2 Ops to Level1, which also solves this issue.

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**